src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,784 @@
+  {
+    QgsVector res = interpolatedValuePrivate( faceIndex, point );
+    if ( isVectorValid( res ) )
+    {
+      mCacheFaceIndex = faceIndex;
+      activeFaceFilter( res, mCacheFaceIndex );
+      return res;
+    }
+  }
+
+  //--> no face found return non valid vector
+  return ( QgsVector( std::numeric_limits<double>::quiet_NaN(), std::numeric_limits<double>::quiet_NaN() ) );
+
+}
+
+QgsMeshVectorValueInterpolatorFromVertex::QgsMeshVectorValueInterpolatorFromVertex( const QgsTriangularMesh &triangularMesh, const QgsMeshDataBlock &datasetVectorValues ):
+  QgsMeshVectorValueInterpolator( triangularMesh, datasetVectorValues )
+{
+
+}
+
+QgsMeshVectorValueInterpolatorFromVertex::QgsMeshVectorValueInterpolatorFromVertex( const QgsTriangularMesh &triangularMesh, const QgsMeshDataBlock &datasetVectorValues, const QgsMeshDataBlock &scalarActiveFaceFlagValues ):
+  QgsMeshVectorValueInterpolator( triangularMesh, datasetVectorValues, scalarActiveFaceFlagValues )
+{
+
+}
+
+QgsVector QgsMeshVectorValueInterpolatorFromVertex::interpolatedValuePrivate( int faceIndex, const QgsPointXY point ) const
+{
+  QgsMeshFace face = mTriangularMesh.triangles().at( faceIndex );
+
+  QgsPoint p1 = mTriangularMesh.vertices().at( face.at( 0 ) );
+  QgsPoint p2 = mTriangularMesh.vertices().at( face.at( 1 ) );
+  QgsPoint p3 = mTriangularMesh.vertices().at( face.at( 2 ) );
+
+  QgsVector v1 = QgsVector( mDatasetValues.value( face.at( 0 ) ).x(),
+                            mDatasetValues.value( face.at( 0 ) ).y() );
+
+  QgsVector v2 = QgsVector( mDatasetValues.value( face.at( 1 ) ).x(),
+                            mDatasetValues.value( face.at( 1 ) ).y() );
+
+  QgsVector v3 = QgsVector( mDatasetValues.value( face.at( 2 ) ).x(),
+                            mDatasetValues.value( face.at( 2 ) ).y() );
+
+
+  return QgsMeshLayerUtils::interpolateVectorFromVerticesData(
+           p1,
+           p2,
+           p3,
+           v1,
+           v2,
+           v3,
+           point );
+
+}
+
+QgsMeshVectorValueInterpolator::QgsMeshVectorValueInterpolator( const QgsTriangularMesh &triangularMesh,
+    const QgsMeshDataBlock &datasetVectorValues ):
+  mTriangularMesh( triangularMesh ),
+  mDatasetValues( datasetVectorValues ),
+  mUseScalarActiveFaceFlagValues( false )
+{
+
+}
+
+QgsMeshVectorValueInterpolator::QgsMeshVectorValueInterpolator( const QgsTriangularMesh &triangularMesh,
+    const QgsMeshDataBlock &datasetVectorValues,
+    const QgsMeshDataBlock &scalarActiveFaceFlagValues ):
+  mTriangularMesh( triangularMesh ),
+  mDatasetValues( datasetVectorValues ),
+  mActiveFaceFlagValues( scalarActiveFaceFlagValues ),
+  mUseScalarActiveFaceFlagValues( true )
+{
+
+}
+
+
+void QgsMeshVectorValueInterpolator::updateCacheFaceIndex( const QgsPointXY &point ) const
+{
+  if ( ! QgsMeshUtils::isInTriangleFace( point, mFaceCache, mTriangularMesh.vertices() ) )
+  {
+    /*the first version of this method uses QgsGeometry::contain to check if the point is in triangles,
```

Comment:
this is probably developer comment, best to remove it for production version and also rename the function you want to use .remove the other one